ksqldb-common/src/main/java/io/confluent/ksql/schema/ksql/LogicalSchema.java

```diff
@@ -236,48 +269,116 @@ public String toString(final FormatOptions formatOptions) {
-    final List<Column> value = byNamespace.get(VALUE);
 
     final ImmutableList.Builder<Column> builder = ImmutableList.builder();
 
-    builder.addAll(key);
+    final ImmutableList<Column> keyColumns = keyColumns(byNamespace);
 
-    int valueIndex = 0;
-    for (final Column c : value) {
-      if (SystemColumns.isSystemColumn(c.name())) {
-        continue;
-      }
+    final ImmutableList<Column> nonPseudoAndKeyCols = nonPseudoAndKeyCols(
+        byNamespace, pseudoColumnVersion, this);
 
-      if (findColumnMatching(withNamespace(Namespace.KEY).and(withName(c.name()))).isPresent()) {
-        continue;
-      }
+    builder.addAll(keyColumns);
+    builder.addAll(nonPseudoAndKeyCols);
+
+    int valueIndex = nonPseudoAndKeyCols.size();
+
+    if (pseudoColumnVersion >= ROWTIME_PSEUDOCOLUMN_VERSION) {
+      builder.add(Column.of(ROWTIME_NAME, ROWTIME_TYPE, VALUE, valueIndex++));
+    }
+
+    if (pseudoColumnVersion >= ROWPARTITION_ROWOFFSET_PSEUDOCOLUMN_VERSION) {
+      builder.add(Column.of(ROWPARTITION_NAME, ROWPARTITION_TYPE, VALUE, valueIndex++));
+      builder.add(Column.of(ROWOFFSET_NAME, ROWOFFSET_TYPE, VALUE, valueIndex++));
+    }
 
+    for (final Column c : key) {
       builder.add(Column.of(c.name(), c.type(), VALUE, valueIndex++));
     }
 
-    if (withPseudoAndKeyColsInValue) {
-      builder.add(Column.of(ROWTIME_NAME, ROWTIME_TYPE, VALUE, valueIndex++));
+    if (windowedKey) {
+      builder.add(
+          Column.of(WINDOWSTART_NAME, WINDOWBOUND_TYPE, VALUE, valueIndex++));
+      builder.add(
+          Column.of(WINDOWEND_NAME, WINDOWBOUND_TYPE, VALUE, valueIndex));
+    }
+
 
-      for (final Column c : key) {
-        builder.add(Column.of(c.name(), c.type(), VALUE, valueIndex++));
+    return new LogicalSchema(builder.build());
+  }
+
+  /**
+   * Rebuilds schema without pseudocolumns or key columns
+   * @return the LogicalSchema created, with the corresponding pseudo and key columns excluded
+   */
+  private LogicalSchema rebuildWithoutPseudoAndKeyColsInValue(final int pseudoColumnVersion) {
+    final ImmutableList.Builder<Column> builder = ImmutableList.builder();
+
+    final ImmutableList<Column> keyColumns = keyColumns(byNamespace());
+
+    final ImmutableList<Column> nonPseudoAndKeyCols = nonPseudoAndKeyCols(
+        byNamespace(), pseudoColumnVersion, this);
+
+    builder.addAll(keyColumns);
+    builder.addAll(nonPseudoAndKeyCols);
+
+    return new LogicalSchema(builder.build());
+  }
+
+  /**
+   * Adds columns, except for key and pseudocolumns, to a provided builder's value schema
+   * and returns the number of added columns
+   * @param builder the builder to be passed in. Columns that don't qualify as key or pseudocolumns
+   *                will be added
+   * @return the number of columns added. This also functions as a "column index", to pass to
+   *                builder.add()
+   * */
+  private static ImmutableList<Column> nonPseudoAndKeyCols(
```

Comment:
Does this method need to be static? Rather than passing in a LogicalSchema to a static method, it'd be nice to have an instance method instead.

Reply:
Let's also update the method name to clarify that the columns being returned are value columns, since that would not be my assumption otherwise as a reader.
```suggestion
  private static ImmutableList<Column> nonPseudoAndKeyColsAsValueCols(
```